plugins/inputs/salesforce/salesforce.go

```diff
@@ -0,0 +1,255 @@
+			return l, err
+		}
+		defer resp.Body.Close()
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return l, fmt.Errorf("Salesforce responded  with unexpected status code %d", resp.StatusCode)
+	}
+
+	l = limits{}
+	err = json.NewDecoder(resp.Body).Decode(&l)
+	return l, err
+}
+
+func (s *Salesforce) getLoginEndpoint() (string, error) {
+	switch s.Environment {
+	case "sandbox":
+		return fmt.Sprintf("https://test.salesforce.com/services/Soap/c/%s/", s.Version), nil
+	case "production":
+		return fmt.Sprintf("https://login.salesforce.com/services/Soap/c/%s/", s.Version), nil
+	default:
+		return "", fmt.Errorf("unknown environment type: %s", s.Environment)
+	}
+}
+
+// Authenticate with Salesfroce
+func (s *Salesforce) login() error {
+	if s.Username == "" || s.Password == "" {
+		return errors.New("missing username or password")
+	}
+
+	body := fmt.Sprintf(`<?xml version="1.0" encoding="utf-8"?>
+		<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
+		  xmlns:urn="urn:enterprise.soap.sforce.com">
+		  <soapenv:Body>
+			<urn:login>
+			  <urn:username>%s</urn:username>
+			  <urn:password>%s%s</urn:password>
+			</urn:login>
+		  </soapenv:Body>
+		</soapenv:Envelope>`,
+		s.Username, s.Password, s.SecurityToken)
+
+	loginEndpoint, err := s.getLoginEndpoint()
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, loginEndpoint, strings.NewReader(body))
+	if err != nil {
+		return err
+	}
+	req.Header.Add("Content-Type", "text/xml")
+	req.Header.Add("SOAPAction", "login")
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	soapFault := struct {
+		Code    string `xml:"Body>Fault>faultcode"`
+		Message string `xml:"Body>Fault>faultstring"`
+	}{}
+
+	err = xml.Unmarshal(respBody, &soapFault)
+	if err != nil {
+		return err
+	}
+
+	if soapFault.Code != "" {
+		return fmt.Errorf("login failed: %s", soapFault.Message)
+	}
+
+	loginResult := struct {
+		ServerURL      string `xml:"Body>loginResponse>result>serverUrl"`
+		SessionID      string `xml:"Body>loginResponse>result>sessionId"`
+		OrganizationID string `xml:"Body>loginResponse>result>userInfo>organizationId"`
+	}{}
+
+	err = xml.Unmarshal(respBody, &loginResult)
+	if err != nil {
+		return err
+	}
+
+	s.SessionID = loginResult.SessionID
+	s.OrganizationID = loginResult.OrganizationID
+	s.ServerURL, err = url.Parse(loginResult.ServerURL)
+
+	return err
+}
+
+func init() {
+	inputs.Add("salesforce", func() telegraf.Input {
+		return NewSalesforce()
+	})
+}
+
+var matchFirstCap = regexp.MustCompile("(.)([A-Z][a-z]+)")
+var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
+
+// convet a string to snake_case
+func toSnakeCase(str string) string {
+	snake := matchFirstCap.ReplaceAllString(str, "${1}_${2}")
+	snake = matchAllCap.ReplaceAllString(snake, "${1}_${2}")
+	return strings.ToLower(snake)
+}
```

Comment:
You can use this function https://github.com/influxdata/telegraf/blob/master/internal/internal.go#L160 for converting to snake case.